resources/newssite/news-next/public/test-invoker.mjs

```diff
@@ -0,0 +1,55 @@
+class TestInvoker {
```

Comment:
I was surprised this is in public/ since you are loading this file via npm (I would have expected it to only exist in dist/). Is this left over or is this how npm works?

Reply:
yeah, so basically I'm doing this:

1. install dependency through [package.json](https://github.com/WebKit/Speedometer/blob/ee3ceb4c4598ae7517d6a403e99cefdb56a38a53/resources/newssite/news-next/package.json#L24) (this copies the two files in my node_modules folder)
2. run a [script](https://github.com/WebKit/Speedometer/blob/ee3ceb4c4598ae7517d6a403e99cefdb56a38a53/resources/newssite/news-next/scripts/prepare.js#L39) to copy both files to the public folder

The reason is that the code needs to execute in the hosted directory when you visit the workload. 
The node_modules folder is not available at that point.

Reply:
if you were to update the `<script>` src in the development source to point to the full node_modules path wouldn't the build step be able to figure it out to get it into dist without the extra script to copy the file? Not sure how the build works exactly but that's what I figured would happen conceptually

Reply:
let me take a look if I can simplify....

Reply:
@bgrins - looking at the code, the files I am referencing in the `script` component are all files created by the news-site-workload:

- benchmark-connector.mjs
- workload-test.mjs
- workload-testing-utils.mjs

The last file imports the test-runner, but since I am using the script component, I have to put the 3 files above in the public folder. This is because of the way Next.js resolves file paths.

Second problem: I can't reference an import from the node_modules folder, since the three files above are in the public folder, which doesn't have access to the node_modules.

Reply:
As a test, I ran `npm create vite@latest` and then installed the entire "speedometer" package rather than one-by-one mjs, then referenced it as `import { TestRunner } from 'speedometer/resources/test-runner.mjs'` and it seems to work in both dev and build (see screenshot). I assume with this approach we could also move workload-testing-utils.mjs into the main resources dir instead of implementing it directly in the workload.

I'm not quite sure how Next `<Script>` elements integrate with this, but presumably they have a way to reference something from an installed package (maybe make the script element reference a local file and then have that file ESM import the relevant helpers). Or maybe I'm missing something?

![Screenshot 2024-11-21 at 10 33 55 AM](https://github.com/user-attachments/assets/c604537a-b6db-4f0b-b220-f7c3856e38e4)

Reply:
#459 shows an example of this

Reply:
we need a solution that works for various frameworks though. 
For Next.js, I can't point the Script component to the node_modules folder... 
for example below won't work

```
<Script src="node_modules/benchmark-connector.mjs"...>
```

Reply:
> #459 shows an example of this

yeah imports are different from static js files though.

Reply:
@bgrins was convincing 😄 .... refactored and introduced a shared folder that's installable.